### ISSUE TITLE
Default RPC methods handler

### ISSUE THREAD

**Issue by xlc** [paritytech/jsonrpsee#984]
How to setup a default RPC method handler to handle all unregistered RPC calls?

**Comment by jsdw** [comment]
This isn't possible at the moment over WS. Via HTTP one could possibly add a tower middleware to catch unknown method names (though it may not be very ergonomic).

We could conceivably support registering a fallback handler to catch such requests though; seems like a decent idea!

**Comment by lexnv** [comment]
Hey, 

Looking at the code, it seems that we treat all methods that are not registered under the same bracket:

https://github.com/paritytech/jsonrpsee/blob/ed9a3ee9953970b4bf0eb7474d58d6d79d0effed/core/src/server/rpc_module.rs#L427-L428

It sounds like a nice ergonomic boost to let users optionally decide how to handle the unregistered methods for sure.

**Comment by xlc** [comment]
Yes this will be super useful. I am working a JSON RPC proxy server (something like this but with up-to-date deps) and would want to forward all unknown RPC calls to upstream.

**Comment by Vid201** [comment]
This would be very helpful.

Is this being worked on? Is there are any working example with tower middleware anywhere?

**Comment by jsdw** [comment]
We have one middleware here that can serve as an example of how to build them:

https://github.com/paritytech/jsonrpsee/blob/master/server/src/middleware/proxy_get_request.rs

And then a couple of examples of how to add middlewares when starting up a server:

https://github.com/paritytech/jsonrpsee/blob/master/examples/examples/http_proxy_middleware.rs
https://github.com/paritytech/jsonrpsee/blob/master/examples/examples/middleware.rs

Beyond that I'd have a look into thw `tower` docs to learn more about how to build middlewares, eg https://github.com/tower-rs/tower/blob/master/guides/building-a-middleware-from-scratch.md.

I don't believe that this has seen any work of late; we've been busy getting the backpressure work finished. We'd happily accept a PR to add something here though (maybe something like a callback that can be registered which fires for all "unknown" requests or something?) :)

**Comment by niklasad1** [comment]
It should be quite to easy to add a configuration to the server that can overwrite the default JSON-RPC behavior i.e, "method not found" when a method is not available on the server. 

Personally, I don't really follow why this feature is needed you could just check if the `JSON-RPC response error code == 32601` in "RpcLogger" and then take some action such as `prometheus metrics` but I suppose you want to intercept the entire call in your that proxy but you will likely run into similar issues for `invalid request`, `invalid params`, `internal error`, `parse error` and similar.

Thus, just having the default handler for just `invalid method name/unregistered methods` makes it a bit weird/inconsistent but still useful. Ideally this would be solved by some middleware instead.

Currently, it's possible to write middleware for this but it's a bit hairy because you would have to parse the response and if it's `JSON-RPC error code 32601` then modify the response for your usecase.

**Comment by Vid201** [comment]
Thank you both @jsdw and @niklasad1 for providing more information! Yeah, I'll go with a custom middleware for now, seems the best option at the moment.

Maybe I'll look at this more in the future and make a PR.

**Comment by xlc** [comment]
middleware doesn’t work for websocket which is the whole point for me to use jsonrpsee. I might as well switch to something else if I am going to write the same tower middleware

**Comment by niklasad1** [comment]
I meant that it would be better introduce some kind of customized middleware for "JSON-RPC calls" not `tower" as it applies only on the HTTP level which doesn't work for Websockets properly.
